src/adr/0018-jmap-new-specs.md

```diff
@@ -0,0 +1,48 @@
+# 18. New JMAP specifications adoption
+
+Date: 2020-02-06
+
+## Status
+
+Proposed
+
+## Context
+
+Historically, James is an early adopter for the JMAP specification, and a first partial implementation was conducted when JMAP was just a draft. But with time, the IETF draft went with radical changes and the community could not keep this implementation up to date with the spec changes.
+
+As off summer 2019, JMAP core ([RFC 8620](https://tools.ietf.org/html/rfc8620)) and JMAP mail ([RFC 8621](https://tools.ietf.org/html/rfc8621)) have been officially published. Thus we should implement these new specifications.
+
+We need to keep in mind though that part of the community actively relies on the actual 'draft' implementation of JMAP existing in James. We should ensure no changes is done to that 'draft' protocol is done while implementing the new one.
+
+## Decision
+
+The proposed approach would be as follow:
+
+* Rename packages `server/protocols/jmap*` and guice packages `server/container/guice/protocols/jmap*` to `jmap-draft`. `JMAPServer` should also be renamed to `JMAPDraftServer` (this has already been contributed [here](https://github.com/apache/james-project/pull/164), thanks to @cketti). No modification in `jmap-draft` integration tests should be conducted afterwards.
+* Create a new `jmap` package, binding it on a new port.
```

Comment:
It's very unusual to use a different port for different versions of a protocol.
Other solutions are version in an HTTP header and version in URL path.
Why should we choose a different port?

Reply:
@aduprat asked something similar in the mail discussion, to which @chibenwa answered : 
```
The proposal to expose the jmap protocol and jmap-draft proptocol on
different port was to avoid "clash" and confusion on other endpoints
(upload, download, authentication).


I tend to find the "full separation" more clear. Having different
port/maven module will also lead to two distinct test suite, and will
ensure we do not break existing implementation with refactorings.
```
https://www.mail-archive.com/server-dev@james.apache.org/msg62084.html

What is wrong with using an other port except for "it's unusual"? Version in the url path would be ok too in my point of view. HTTP header... could you give me an example for that one?

Reply:
https://dzone.com/articles/best-practice-api-versioning-for-http-rest-interfa or https://www.baeldung.com/rest-versioning

We had the discussion previously with @jeantil too

What must not change is the testsuite, the jmap-draft can change if required.

BTW, the change could be located at a single place and the only change required would be the `Accept` header in the rest-assured RequestSpecification object.

> What is wrong with using an other port except for "it's unusual"?

Also, we should use unusual solutions only when we can explain why because it breaks the "Principle of least astonishment" that is a valuable principle.

However, a very good reason to use another port would be we want to change the server technology and we don't want to port jmap-draft to this new technology.

Reply:
A quick note: both links suggest using vnd that is custom mime types which may or may not be registered. They both overlook the fact that the HTTP specification allows for extensions on the mime-type, not only the weight which is treated specially `q=`. 

Therefore an alternative to vnds is to use standard mime types or a single unversionned vnd and add an accept-extension for the version. For example : 
```
Accept: application/json;v=2.3.1
``` 
or 
```
Accept: application/vnd.jmap+json;v=2.3.1
``` 
See https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html for more details about accept-extensions.

Reply:
Having things served on a same port will ease rolling frontend adoption (eg: use new spec for listing mailboxes, rely on the old one for messages)

Thanks @jeantil for pointing us to best practices :-)

Reply:
that's interesting, but here we are not versionning an API, we are progressively removing an implementation of a never-really-existed-version of a specification by the only version currently existing. So do we want to have to add a version to each of our future http requests to handle the difference with a protocol we will probably remove?

At first sight I would have say that it is not needed, but in fact JMAP is already planning a potential V2. So maybe we can already anticipate this and enforce the version in every http request.

Reply:
`Accept: application/json;jmapVersion=draft` vs `Accept: application/json;jmapVersion=rfc-8620`

If jmapVersion is ommited we default to draft, and to rfc-8620 when draft is deprecated or removed?